File: src/main/java/me/lokka30/levelledmobs/events/MobPostLevelEvent.java
/*
 * This file is Copyright (c) 2020-2022 lokka30.
 * This file is/was present in the LevelledMobs resource.
 * Repository: <https://github.com/lokka30/LevelledMobs>
 * Use of this source code is governed by the GNU GPL v3.0
 * license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.events;

import org.bukkit.event.Cancellable;
import org.bukkit.event.Event;
import org.bukkit.event.HandlerList;
import org.jetbrains.annotations.NotNull;

/**
 * @author lokka30
 * @see Event
 * @see Cancellable
 * @see MobPreLevelEvent
 * @since v4.0.0
 * This event is fired by LevelledMobs when a
 * mob has become Levelled. It is not Cancellable
 * since MobPreLevelEvent allows plugins to do such.
 */
public class MobPostLevelEvent extends Event {

    /*
    TODO
        lokka30: Complete event class.
     */

    @NotNull
    @Override
    public HandlerList getHandlers() {
        return null;
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/events/MobPostLevelEvent.kt
package io.github.arcaneplugins.levelledmobs.events

import io.github.arcaneplugins.levelledmobs.result.AdditionalLevelInformation
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper
import org.bukkit.Bukkit
import org.bukkit.entity.LivingEntity
import org.bukkit.event.Event
import org.bukkit.event.HandlerList

/**
 * This event is fired *after* a mob is levelled. Other plugins can cancel this event.
 *
 * @author lokka30
 * @since 2.5.0
 */
class MobPostLevelEvent(
    val lmEntity: LivingEntityWrapper,
    val levelCause: LevelCause,
    val additionalInformation: MutableSet<AdditionalLevelInformation?>
) : Event(!Bukkit.isPrimaryThread()) {
    companion object {
        private val HANDLERS = HandlerList()

        fun getHandlerList(): HandlerList {
            return HANDLERS
        }
    }

    override fun getHandlers(): HandlerList {
        return HANDLERS
    }

    /**
     * When a mob is levelled, the following enum is used to allow plugins to find the cause of the
     * mob being levelled.
     *
     *
     * NORMAL: Spawned naturally, by a spawn egg, etc. CHANGED_LEVEL: When an existing levelled mob
     * has its level changed.
     */
    enum class LevelCause {
        NORMAL,
        CHANGED_LEVEL,
        SUMMONED
    }

    val entity: LivingEntity
        get() = lmEntity.livingEntity

    val level: Int
        get() = lmEntity.getMobLevel()
}